x/evm/keeper/msg_server.go

```diff
@@ -0,0 +1,187 @@
+package keeper
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"math/big"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core"
+	ethtypes "github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/core/vm"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/sei-protocol/sei-chain/x/evm/state"
+	"github.com/sei-protocol/sei-chain/x/evm/types"
+	tmtypes "github.com/tendermint/tendermint/types"
+)
+
+type msgServer struct {
+	Keeper
+}
+
+// NewMsgServerImpl returns an implementation of the MsgServer interface
+// for the provided Keeper.
+func NewMsgServerImpl(keeper Keeper) types.MsgServer {
+	return &msgServer{Keeper: keeper}
+}
+
+var _ types.MsgServer = msgServer{}
+
+func (server msgServer) EVMTransaction(goCtx context.Context, msg *types.MsgEVMTransaction) (*types.MsgEVMTransactionResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	stateDB := state.NewStateDBImpl(ctx, &server)
+	tx, _ := msg.AsTransaction()
+	ctx, gp := server.getGasPool(ctx)
+	emsg, err := server.getEVMMessage(ctx, tx)
+	if err != nil {
+		return &types.MsgEVMTransactionResponse{}, err
+	}
+
+	res, err := server.applyEVMMessage(ctx, emsg, stateDB, gp)
+	if err != nil {
+		return &types.MsgEVMTransactionResponse{}, err
+	}
+
+	if err := server.writeReceipt(ctx, tx, emsg, res); err != nil {
+		return &types.MsgEVMTransactionResponse{}, err
+	}
+
+	if err := stateDB.Finalize(); err != nil {
+		return &types.MsgEVMTransactionResponse{}, err
+	}
+
+	errStr := ""
+	if res.Err != nil {
+		errStr = res.Err.Error()
+	}
+
+	return &types.MsgEVMTransactionResponse{
+		GasUsed:    res.UsedGas,
+		ReturnData: res.ReturnData,
+		VmError:    errStr,
+		Hash:       tx.Hash().Hex(),
+	}, nil
+}
+
+func (server msgServer) getGasPool(ctx sdk.Context) (sdk.Context, core.GasPool) {
+	if ctx.BlockGasMeter() == nil {
+		ctx = ctx.WithBlockGasMeter(sdk.NewInfiniteGasMeter())
+	}
+	if ctx.BlockGasMeter().Limit() == 0 {
+		// infinite gas meter
+		return ctx, math.MaxUint64
+	} else {
+		return ctx, core.GasPool(ctx.BlockGasMeter().Limit() - ctx.BlockGasMeter().GasConsumedToLimit())
+	}
+}
+
+func (server msgServer) getEVMMessage(ctx sdk.Context, tx *ethtypes.Transaction) (*core.Message, error) {
+	cfg := server.GetChainConfig(ctx).EthereumConfig(server.ChainID())
+	signer := ethtypes.MakeSigner(cfg, big.NewInt(ctx.BlockHeight()), uint64(ctx.BlockTime().Unix()))
+	return core.TransactionToMessage(tx, signer, server.getBaseFee())
+}
+
+func (server msgServer) applyEVMMessage(ctx sdk.Context, msg *core.Message, stateDB vm.StateDB, gp core.GasPool) (*core.ExecutionResult, error) {
+	coinbase, err := server.GetFeeCollectorAddress(ctx)
```

Comment:
why is it coinbase specific?

Reply:
`coinbase` here is an Ethereum term that's basically equivalent to `feeCollector` in Sei, not the exchange.